src/docs/sphinx/developerGuide/Contributing/CodeRules.rst

```diff
@@ -0,0 +1,1440 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When modifying performance-critical code (kernels, assembly loops, solvers):
+
+1. **Use Caliper profiling** on representative test cases before changes
+2. **Measure performance** after changes
+3. **Document performance impact** in pull request description
+
+.. dropdown:: Example: Performance profiling workflow
+   :icon: code
+
+   .. code-block:: bash
+
+      # Profile before changes
+      geos -i test_case.xml --caliper-output=baseline.cali
+      
+      # Make code changes
+      
+      # Profile after changes
+      geos -i test_case.xml --caliper-output=modified.cali
+      
+      # Compare results
+      cali-query -q "SELECT time.duration WHERE annotation='myKernel'" baseline.cali
+      cali-query -q "SELECT time.duration WHERE annotation='myKernel'" modified.cali
+
+Caliper Integration
+^^^^^^^^^^^^^^^^^^^^
+
+Use ``GEOS_MARK_FUNCTION`` and ``Timer`` for performance tracking for the main computation functions / scopes.
+
+.. dropdown:: Example: Performance instrumentation
+   :icon: code
+
+   .. code-block:: c++
+
+      void PhysicsSolverBase::solverStep( ... )
+      {
+        GEOS_MARK_FUNCTION; // Caliper tracking for entire function
+
+        {
+          GEOS_CALIPER_MARK_SCOPE("assemble system");
+          Timer timer( m_timers.get_inserted( "assembly" ) );
+          assembleSystem();
+        }
+
+        {
+          GEOS_CALIPER_MARK_SCOPE("linear solve");
+          Timer timer( m_timers.get_inserted( "linear solver" ) );
+          linearSolve();
+        }
+      }
+
+Speed Optimization Rules
+-----------------------------
+
+- **Hoist Loop Invariants**, move computations that don't change during iterations outside the loop.
+- When it does not critically affect the code architecture and clarity, **Fuse multiple related kernels to reduce memory traffic and launch overhead** (i.e., statistics kernels process all physics field at once).
+- **Optimize Memory Access for Cache and Coalescing**. Access memory sequentially and ensure coalesced access, especially on GPUs.
+- **Minimize Host-Device Transfers**. Keep data on the appropriate memory space and minimize transfers.
+
+Memory Management Rules
+---------------------------------
+
+Generalities
+^^^^^^^^^^^^^^^^^^^^
+
+- **Minimize dynamic memory allocation** as much as possible, particularly in performance-critical code,
+- For frequently allocated/deallocated objects, **consider memory pools**,
+- For containers with known size, **reserve capacity upfront**.
+
+Be Const / Constexpr
+^^^^^^^^^^^^^^^^^^^^
+
+**Use ``const`` and ``constexpr`` when possible**, enabling:
+
+- **Compiler optimization,** enables better code generation by giving constraints to the code,
+- **Code safety,** prevents accidental modification for constant contexts,
+- **Show code intention,** make code more readable.
+
+Also, **Mark methods const if the method is not designed to modify** the object state.
+
```

Comment:
Same as above, it sounds like very high level advises, maybe better to relocate in another doc ?